### ISSUE TITLE
Sensitive Data: Mitigate against password guessing attacks

### ISSUE THREAD

**Issue by pdurbin** [IQSS/dataverse#3153]
http://policy.security.harvard.edu/sb2-password-guessing says the following as of 2016-06-01:

>  Password guessing
> 
> SB2: Servers or applications must implement a mechanism that inhibits password guessing attacks on user accounts if the server or application does its own authentication.
> 
> How to Comply
> 
> Block excessive logins
> 
>  Block user from logging in for a period of time after no more than 10 successive invalid login attempts.

Dataverse **does** do its own authentication in the case of local/builtin accounts so we need to meet this requirement to be able to host level 3 data as defined at http://policy.security.harvard.edu/level-3

When Shibboleth accounts are used, we're trusting that the Identity Provider (IdP) is mitigating against password guessing attacks.

We should decide what "block user from logging in for a period of time" means. My initial thought would that this would be disabling the account for an hour or whatever (configurable) but this is not yet supported (#2419) and from reading https://www.owasp.org/index.php/Blocking_Brute_Force_Attacks I'm realizing that there's the potential for abuse. An attacker who knows your username or email address for your local/builtin account would be able to lock you out of your account temporarily simply by guessing random passwords 10 times. That OWASP page suggests blocking the IP of the attacker, which may be problematic if many users are behind one IP address. It also suggests a CAPTCHA but I don't think that would meet the "block user" requirement. Their suggestion to introduce a delay against the IP of the attacker reminds me strongly of #1339 which is about adding rate limiting to the API.

**Comment by pdurbin** [comment]
I posted a link to this issue to https://groups.google.com/d/msg/dataverse-dev/hxnA_Wumo_M/J41aNceCBAAJ and http://irclog.iq.harvard.edu/dataverse/2016-06-08#i_36462 and http://irclog.perlgeek.de/crimsonfu/2016-06-09 for discussion and I like the idea from @pameyer and @bpeisenbraun of what seems to be called "exponential backoff". The screenshot below is from https://devcentral.f5.com/articles/implementing-the-exponential-backoff-algorithm-to-thwart-dictionary-attacks and shows how the lockout time increases with each subsequent failed login attempt:

![screen shot 2016-06-09 at 9 28 40 am](https://cloud.githubusercontent.com/assets/21006/15931198/b121e832-2e24-11e6-96df-5e5cb52bd826.png)

If we happened to be using Ruby and Rack, we could take advantage of the exponential backoff feature of rack-attack: https://github.com/kickstarter/rack-attack/wiki/Advanced-Configuration#exponential-backoff

I suppose the thing we would be locking out is an IP address. I'm actually not sure that we need to implement account disabling in #2419 first. The requirement is to block a user from logging in, which is different from disabling an account.

**Comment by pdurbin** [comment]
Obviously, we'd like to avoid the scenario at https://twitter.com/esttorhe/status/743623557352022020 in the screenshot below:

![screen shot 2016-06-17 at 9 48 27 am](https://cloud.githubusercontent.com/assets/21006/16152607/bc7050c8-3470-11e6-8e46-0c912ea5aca8.png)

**Comment by pdurbin** [comment]
This morning @djbrooke and I met to discuss this issue. I presented two approaches:

- **Approach 1: Lock account**. After 10 (or fewer, configurable) successive invalid logins, lock the account (#2419) for a period of time (configurable).
- **Approach 2: Block IP**. After 10 (or fewer, configurable) successive invalid logins, block the IP address for a period of time (configurable).

@djbrooke indicated that I should go ahead work on #2419 so that locking an account is possible and then implement Approach 1: Lock account. So this issue is dependent on that one.

As I indicated above, locking the account was my initial thought anyway (perhaps because of how the requirement is written) and @landreev seems to agree with this approach. Yesterday I wrote to [abdc-security](http://www.abcd.harvard.edu) and a friend [reacted](http://irclog.perlgeek.de/crimsonfu/2016-06-27) with, "you are overthinking your password lockout problem... N bad passwords in Y interval locks out the account for 24 hours."

I've been discussing Approach 2 (Block IP) with @whorka and he was gracious enough to show me the code at https://accounts.hmdc.harvard.edu which uses this approach but again, we'll focus on account locking instead (Approach 1).

**Comment by michbarsinai** [comment]
Important heads-up. Thanks, Phil!

My comments:

1. Disabling account is very different from blocking a login. For example, if someone tries to guess user X's password, while user X is logged in from another device or is using the API via a script, X's legitimate usage would be stopped. Moreover - if someone has a password and tries to guess the username, this would go under the radar of account disabling.
2. Blocking an IP would probably not work either, especially with our target audience, as institutions tend to use a few IP addresses for outside access.
3. Also - blocking and tracking done on the server-side add overhead to the server, sometime for each HTTP request.

So - I assume it's best not to allow people to automatically guess password, which is what CAPTCHA does. We could use reCAPTCHA (or similar project, as reCAPTCHA was acquired by Google) to help digitize books while we're at it. So maybe present a CAPTCHA after 3 failed tries from same IP address in a 3-4 minute period. That's still added some burden on the server, but we could use an in-memory asynchronous map, and not have to go to the DB again. Also - this would effect only the login screen.

**Comment by pdurbin** [comment]
@michbarsinai thanks for your thoughts on all this. I'm not sure a CAPTCHA fulfills the requirement of "Block user from logging in for a period of time after no more than 10 successive invalid login attempts."

As I mentioned at https://github.com/IQSS/dataverse/issues/3153#issuecomment-229085993 I'm focusing on locking the account for now. Here's a screenshot to get @mheppler @eaquigley and @mcrosas thinking about what the UI/UX should be:

![log_in_-_root_dataverse_-_2016-06-29_16 03 32](https://cloud.githubusercontent.com/assets/21006/16467071/ad76d914-3e13-11e6-919d-cfcca0ed2d98.png)

I'm also adding logic in EjbDataverseEngine to prevent Commands from being executed when an account is locked. Once I have something worth demo'ing I'll push a war file to a test server.

**Comment by michbarsinai** [comment]
But any disaffected student could lock his professor out this way... maybe the rule needs updating or refinement.

What does Harvard's SHIB/HarvardKey servers do?

**Comment by djbrooke** [comment]
Notes from 7/7 meeting:

- 10 tries before an account is locked
- We need better error message visibility
- The lockout period should be configurable
- Open question - does the superuser have the ability to unlock an account?

**Comment by pdurbin** [comment]
The code I demoed to @djbrooke @mheppler @eaquigley @scolapasta @kcondon and @bsilverstein95 on 2016-07-05 was 6852a96 from https://github.com/pdurbin/dataverse/tree/3153-passwordguess

To echo what @djbrooke wrote, @eaquigley felt that the change from "wrong password" to "locked" was too subtle.

@eaquigley also asked that I test to make sure that resetting one's password doesn't unlock the account. It shouldn't but I'll be happy to test this.

@scolapasta opined that perhaps when the account becomes locked the user should be forced to reset their password. I'm not particularly fond of this idea but I wanted to at least mention it.

In the code I demo'ed a superuser **can** already unlock an account via API: https://github.com/IQSS/dataverse/blob/6852a96d8a6efe6208e94b95cd37dea01a46a5d9/src/main/java/edu/harvard/iq/dataverse/api/Admin.java#L309 . There are already tests around this functionality. I guess I could rip out this feature if we really shouldn't have it.

We also discussed the user story of a user locking themselves out. They are told to contact Support. What should Support do? Should they say, "Just wait until your account is unlocked." Should a superuser have a UI for seeing which accounts are locked? In the harvesting branch a [dashboard](https://github.com/IQSS/dataverse/blob/822e9e6cdbdb78fbf212640fcc32ee5dd9253058/src/main/webapp/dashboard.xhtml) is being developed and #2211 is related.

**Comment by pdurbin** [comment]
I just pushed some code to a branch at https://github.com/IQSS/dataverse/tree/3153-2419-lockuser which is more or less a cleaned up version of what I demo'ed on 2016-07-05. In order to run this code without dropping your database, you need to run the SQL script at https://github.com/IQSS/dataverse/blob/3153-2419-lockuser/scripts/database/upgrades/lockuser.sql

**Comment by pdurbin** [comment]
Today @mheppler and I looked at the latest code in the [3153-2419-lockuser](https://github.com/IQSS/dataverse/tree/3153-2419-lockuser) branch (v. 4.4 build 3153-2419-lockuser-4f05d67 deployed to https://dev1.dataverse.org ).

Out of the box the behavior is "three strikes and you're out" for 60 minutes but both of these values are configurable.

Mike will make any changes to the branch as he sees fit so I'm co-assigning this issue to him.

**Comment by pdurbin** [comment]
Per @djbrooke the Sensitive Data project is being deferred until Dataverse 5.0 so I'm unassigning myself since I like having https://github.com/IQSS/dataverse/issues/assigned/pdurbin reflect what I'm actually working on. I'll also add the 5.0 milestone.

**Comment by pdurbin** [comment]
Sensitive data work isn't happening until Dataverse 5.0. We'll revisit this then. Closing.

**Comment by TaniaSchlatter** [comment]
We use the random math question in a few places – email support, contact, etc. (see screenshot). 
Can we use this after 3 failed log in tries?
<img width="616" alt="screen shot 2017-08-09 at 12 11 10 pm" src="https://user-images.githubusercontent.com/15386424/29131934-78b3bf34-7cfc-11e7-8382-a248a3e37ffa.png">

**Comment by oscardssmith** [comment]
I like this idea. It isn't as annoying as a capcha. It won't prevent specialized attacks, but along with https://www.owasp.org/index.php/Slow_Down_Online_Guessing_Attacks_with_Device_Cookies, which we should do anyway, it would make attacks very difficult without punishing users.

**Comment by mheppler** [comment]
**UI IMPACT**

**Log In**
- [ ] add "Log In Disabled" error msg

![login-locked](https://user-images.githubusercontent.com/687227/29377253-4360aa1c-8289-11e7-89bd-73671500a8ae.gif)

**Comment by oscardssmith** [comment]
looks good. Do we also want to grey out the log in button?

**Comment by pdurbin** [comment]
@matthew-a-dunlap as I mentioned, as part of this effort, we might want to reconsider the feature we added in #1818 to allow users to look up their API token using their username and password. It looks like bf59f53 contains most of the implementation and docs.

**Comment by oscardssmith** [comment]
What criteria were decided on for when to lock an account?

**Comment by matthew-a-dunlap** [comment]
@oscardssmith Nothing has been decided at this point. I just emailed Christian Hamer at Harvard Security to ask for some clarification, and we will go ahead with discussions on the topic in parallel.

**Comment by matthew-a-dunlap** [comment]
Here is a summary of the discussion I've seen so far, in hopes that it will make it easier to have a group discussion:

What do we actually need?
* Foremost, to comply with http://policy.security.harvard.edu/sb2-password-guessing
    * SB2: Servers or applications must implement a mechanism that inhibits password guessing attacks on user accounts if the server or application does its own authentication.
    * Block excessive logins: Block user from logging in for a period of time after no more than 10 successive invalid login attempts.
* Second, are there other security changes we need to improve security in this area to a reasonable level?
    * Locking down our api endpoint for getting an api token with username/password

Questions:
* What is the minimum bar we need to fulfill for Harvard Security?
    * I have sent an email to Christian Hamer asking about whether solely a captcha can fulfill the requirements, or if not what more we would need.
* If we need to engage in some sort of lockout or timer on subsequent log-ins, how do we go about that?
    * Blocking based on IP?
    * Blocking based on account?
    * How do we calculate the pause in time before logging in?
    * Do we lock out an account completely at some point and require admin attention?
    * Can we utilize our cookies to be more intelligent about how we lock out an account?

**Comment by oscardssmith** [comment]
Blocking based on IP will not work because botnets and or tor make attacks from many IP's easy, and because some institutions share few IP addresses, so one person on that network would lock everyone else out.
Blocking based on account will not work, because then one attacker can completely shut down a dataverse by guessing 10 fake passwords for every account once per time period, which is not computationally infeasible
The only solution that prevents account locking attacks, while providing security is a cookie based system where each known good user has a separate guess pool from anyone else, so that even if someone's account is being attacked, that user can still log in from a machine they have used before. The easiest way to calculate the pause in time probably would be to store the date of a lock, and block login attempts until the date changes. Similarly, dates can be put in the known good user cookies, to prevent the machine of a previously known good user from being able to brute force that account

**Comment by whorka** [comment]
The botnet and DoS issues are both concerns that we had to address when 
designing the HMDC Account Self-Service web app.

The way I solved the issue there was to provide a password reset 
mechanism within the app, implemented in such a way so that when a user 
resets their password, it also removes their account lockout.

  -Bill

**Comment by oscardssmith** [comment]
what happens if attackers request password resets?

**Comment by whorka** [comment]
When a password reset is requested, it sends an email with a password 
reset token to the account-holder's email address (regardless of who 
requested it).

  -Bill

**Comment by matthew-a-dunlap** [comment]
I sent an email to Christian Hamer (Chief of Harvard Information Security) to ask whether requiring a captcha after first login would fulfill Harvard's requirements. His response:

> Thanks for asking, Matthew.  As the how-to suggests, most such mechanisms are really about rate limiting (considering automated attempts as one type of threat).  I think the solution you propose is very much in keeping with that approach and addresses the threat of automated password guessing (and makes human password guessing much more challenging).  So that seems perfectly reasonable to me.

In our morning meeting we agreed a captcha would be the best option at this point as we have not had majors issues of accounts being compromised. If in the future a captcha does not fill our needs, solutions along the lines of @oscardssmith, @whorka et al's discussion will be readdressed.

**Comment by pdurbin** [comment]
I pushed an experiment to https://github.com/pdurbin/dataverse/commit/bad1b7f7ceb8e0242085f7b7e1bd2ac0820fa37b but I think it would be extremely messy to reuse the existing SendFeedbackDialog.java/contactFormFragment.xhtml code since it's so oriented toward sending email. @TaniaSchlatter wants a pop up, which is fine but it will be smaller, maybe something like this:

<img width="682" alt="screen shot 2017-10-10 at 4 36 51 pm" src="https://user-images.githubusercontent.com/21006/31409267-4396313e-add9-11e7-90b5-450b23a1656f.png">

**Comment by oscardssmith** [comment]
I'd like to quickly point out that it would be a significantly less trivial captca if the problem was a picture (unobscured), than if it is just text.

**Comment by pdurbin** [comment]
I just made pull request #4193 and am putting this issue into Code Review at https://waffle.io/IQSS/dataverse

I believe @mheppler and I have delivered the non-popup math challenge after three failed login attempts as specified but I'd like to remind everyone the it's also possible to conduct password guessing attack via API like this:

`curl 'http://localhost:8080/api/builtin-users/dataverseAdmin/api-token?password=dictionaryAttack1'`

`curl 'http://localhost:8080/api/builtin-users/dataverseAdmin/api-token?password=dictionaryAttack2'`

etc.

This was implemented in #1818 and is documented in http://guides.dataverse.org/en/4.8/api/native-api.html#retrieving-the-api-token-of-a-builtin-user but unless we hear strong objections from @astrofrog (who opened that issue) or anyone at @CenterForOpenScience I think we should consider disabling this API out of the box and providing a mechanism to turn it back on if Dataverse installations really want it and more API tests. Thoughts?

**Comment by pdurbin** [comment]
It wasn't hard for @mheppler to convince me to go ahead disable to ability to look up one's API token via API using your username and password, which I did in fd743e2 and added a new database setting to revert to the old behavior.

**Comment by eunices** [comment]
> **UI IMPACT**
> 
> **Log In**
> 
> * [ ]  add "Log In Disabled" error msg (third frame of the animated GIF below...)
> 
> ![login-locked](https://user-images.githubusercontent.com/687227/29377442-d6c891d4-8289-11e7-99be-6f2d8ac86fe4.gif)

Hi devs, to check, will this feature be implemented anytime soon?

**Comment by pdurbin** [comment]
@eunices thanks for your interest in this. In the future, please create a new issue because we don't always notice comments on old, closed issues.

It looks like we never implemented this exact "Log In Disabled" message as part of dealing with password guessing attacks. (We do show a "Sorry, your account has been deactivated" message if a superuser has deactivated your account (pull request #7629) but these are very different things.) Instead, if the user enters the password wrong 3 times, we start showing a math problem. This is to slow the user down (mitigate against password guessing attacks). Users cannot lock themselves out. Again, please feel free to open a new issue if you're interested in different behavior.

**Comment by eunices** [comment]
@pdurbin will create a new issue if needed.

Got it, we'll raise a new issue if interested in different behaviour.
